pkg/compose/secrets.go

```diff
@@ -94,47 +114,135 @@ func (s *composeService) injectConfigs(ctx context.Context, project *types.Proje
+	valueAsBytes := []byte(value)
+	header := &tar.Header{
+		Name:    config.Target,
+		Size:    int64(len(valueAsBytes)),
+		Mode:    mode,
+		ModTime: time.Now(),
+		Uid:     uid,
+		Gid:     gid,
+	}
+	err = tarWriter.WriteHeader(header)
+	if err != nil {
+		return bytes.Buffer{}, err
+	}
+	_, err = tarWriter.Write(valueAsBytes)
+	if err != nil {
+		return bytes.Buffer{}, err
+	}
+	err = tarWriter.Close()
+	return b, err
+}
+
+func createTarArchiveOf(path string, config types.FileReferenceConfig) (bytes.Buffer, error) {
+	// need to treat files and directories differently
+	fi, err := os.Stat(path)
+	if err != nil {
+		return bytes.Buffer{}, err
+	}
+
+	// if path is not directory, try to treat it as a file by reading its value
+	if !fi.IsDir() {
+		buf, err := os.ReadFile(path)
+		if err == nil {
+			return createTarredFileOf(string(buf), config)
+		}
+	}
+
+	mode, uid, gid, err := makeTarFileEntryParams(config)
+	if err != nil {
+		return bytes.Buffer{}, fmt.Errorf("failed parsing target file parameters")
+	}
+
+	subdir := os.DirFS(path)
+	b := bytes.Buffer{}
+	tarWriter := tar.NewWriter(&b)
+
+	// build the tar by walking instead of using archive/tar.Writer.AddFS to be able to adjust mode, gid and uid
+	err = fs.WalkDir(subdir, ".", func(filePath string, d fs.DirEntry, err error) error {
+		header := &tar.Header{
+			Name:    filepath.Join(config.Target, filePath),
+			Mode:    mode,
+			ModTime: time.Now(),
+			Uid:     uid,
+			Gid:     gid,
+		}
+
+		if d.IsDir() {
+			// tar requires that directory headers ends with a slash
+			header.Name = header.Name + "/"
+			err = tarWriter.WriteHeader(header)
+			if err != nil {
+				return fmt.Errorf("failed writing tar header of directory %v while walking diretory structure, error was: %w", filePath, err)
+			}
+		} else {
+			f, err := subdir.Open(filePath)
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+
+			valueAsBytes, err := io.ReadAll(f)
+			if err != nil {
+				return fmt.Errorf("failed reading file %v for to send to container, error was: %w", filePath, err)
+			}
+
+			header.Size = int64(len(valueAsBytes))
+			err = tarWriter.WriteHeader(header)
+			if err != nil {
+				return fmt.Errorf("failed writing tar header for file %v while walking diretory structure, error was: %w", filePath, err)
+			}
+
+			_, err = tarWriter.Write(valueAsBytes)
+			if err != nil {
+				return fmt.Errorf("failed writing file content of %v into tar archive while walking directory structure, error was: %w", filePath, err)
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return bytes.Buffer{}, fmt.Errorf("failed building tar archive while walking config directory structure, error was: %w", err)
+	}
+
+	err = tarWriter.Close()
+	if err != nil {
+		return bytes.Buffer{}, fmt.Errorf("failed closing tar archive after writing, error was: %w", err)
+	}
+
+	return b, err
+}
+
+func makeTarFileEntryParams(config types.FileReferenceConfig) (mode int64, uid, gid int, err error) {
+	mode = 0o444
```

Comment:
:question: should the default mode be different for secrets compared to configs?

Reply:
I think it would make sense to remove at least public read (`0o440`)